lib/charms/mysql/v0/async_replication.py

```diff
@@ -0,0 +1,524 @@
+
+
+class MySQLAsyncReplication(Object):
+    """MySQL async replication base class."""
+
+    def __init__(self, charm: "MySQLOperatorCharm", relation_name: str):
+        super().__init__(charm, relation_name)
+        self._charm = charm
+
+        # promotion action
+        self.framework.observe(
+            self._charm.on.promote_standby_cluster_action, self._on_promote_standby_cluster
+        )
+
+        # relation broken is observed on all units
+        self.framework.observe(
+            self._charm.on[PRIMARY_RELATION].relation_broken, self.on_async_relation_broken
+        )
+        self.framework.observe(
+            self._charm.on[REPLICA_RELATION].relation_broken, self.on_async_relation_broken
+        )
+
+    @property
+    @lru_cache
+    def role(self) -> ClusterSetInstanceState:
+        """Current cluster set role of the unit, after the relation is established."""
+        if is_replica := self._charm._mysql.is_cluster_replica():
+            cluster_role = "replica"
+        else:
+            cluster_role = "primary" if is_replica is False else "unset"
+
+        _, instance_role = self._charm._mysql.get_member_state()
+
+        if self.model.get_relation(REPLICA_RELATION):
+            relation_side = "replica"
+        else:
+            relation_side = "primary"
+
+        return ClusterSetInstanceState(cluster_role, instance_role, relation_side)
+
+    def get_remote_relation_data(self, relation: Relation) -> Optional[RelationDataContent]:
+        """Remote data."""
+        if not relation.app:
+            return
+        return relation.data[relation.app]
+
+    def _on_promote_standby_cluster(self, event: ActionEvent) -> None:
+        """Promote a standby cluster to primary."""
+        if not self._charm.unit.is_leader():
+            event.fail("Only the leader unit can promote a standby cluster")
+            return
+
+        if not self._charm._mysql.is_cluster_replica():
+            event.fail("Only a standby cluster can be promoted")
+            return
+
+        cluster_set_name = event.params.get("cluster-set-name")
+        if (
+            not cluster_set_name
+            or cluster_set_name != self._charm.app_peer_data["cluster-set-domain-name"]
+        ):
+            event.fail("Invalid cluster set name")
+            return
+
+        # promote cluster to primary
+        cluster_name = self._charm.app_peer_data["cluster-name"]
+        force = event.params.get("force", False)
+
+        try:
+            self._charm._mysql.promote_cluster_to_primary(cluster_name, force)
+            event.set_results({"message": f"Cluster {cluster_name} promoted to primary"})
+        except MySQLPromoteClusterToPrimaryError:
+            logger.exception("Failed to promote cluster to primary")
+            event.fail("Failed to promote cluster to primary")
+
+    def on_async_relation_broken(self, event):  # noqa: C901
+        """Handle the async relation being broken from either side."""
```

Comment:
good point - nothing, just because the handler that deals with the unit being departed will take care of the instance before this hook is handled. In fact, this handler will never have a change to run